presto-matching/src/main/java/com/facebook/presto/matching/v2/Captures.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.matching.v2;
+
+import java.util.NoSuchElementException;
+
+public class Captures
+{
+    private static final Captures NIL = new Captures(null, null, null);
+
+    private final Capture<?> capture;
+    private final Object value;
+    private final Captures tail;
```

Comment:
Why is it better than `(Linked/Immutable)List<Capture>? Maybe lets extract this concept of list as some class?